src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs

```diff
@@ -101,8 +121,22 @@ public override void Abort(ConnectionAbortedException abortReason)
         }
 
         // Only called after connection middleware is complete which means the ConnectionClosed token has fired.
-        public void Dispose()
+        public override async ValueTask DisposeAsync()
         {
+            if (_task != null)
+            {
+                // TODO: Make this timeout configurable, this gives the task time to flush the data to the socket
+                // before timing out
+                var result = await Task.WhenAny(_task, Task.Delay(5000));
```

Comment:
This doesn't *have* to timeout, the previous logic just disposed and assumed timeout logic would be handled by the upper application layers. It could just await the task without aborting.